File: src/rubik/shufflers/tests/shufflers.cpp
#include "testing/testing.h"
#include "rubik/shapes/volume.hpp"
#include "rubik/shapes/cube.hpp"
#include "rubik/shufflers/shufflers.hpp"


template<typename S>
class ShapeShufflerTest : public testing::Test {
};

using ShapeTypes = ::testing::Types<Volume, Cube>;
TYPED_TEST_SUITE(ShapeShufflerTest, ShapeTypes);

TYPED_TEST(ShapeShufflerTest, trivial_difficulty) {
    TypeParam shape;
    auto turns = 0;
    auto seed = 0;
    auto [shuffled, shuffle_path] = specific(shape, turns, seed = 0);
    ASSERT_EQ(shuffled, shape);
    ASSERT_EQ(shuffle_path.size(), turns);
}

TYPED_TEST(ShapeShufflerTest, easy_difficulty) {
    TypeParam shape;
    auto seed = 0;
    for (auto turns: {1, 2}) {
        auto [shuffled, shuffle_path] = specific(shape, turns, seed);
        ASSERT_NE(shuffled, shape);
        ASSERT_EQ(shuffle_path.size(), turns);
    }
}

TYPED_TEST(ShapeShufflerTest, moderate_difficulty) {
    TypeParam shape;
    auto seed = 0;
    for (auto turns = 1; turns < 5; turns++) {
        auto [shuffled, shuffle_path] = specific(shape, turns, seed);
        ASSERT_NE(shuffled, shape);
        ASSERT_EQ(shuffle_path.size(), turns);
    }
}

TYPED_TEST(ShapeShufflerTest, infeasible_moderate_difficulty) {
    GTEST_SKIP(); /* This takes too long. */
    TypeParam shape;
    auto turns = 50;
    auto seed = 0;
    ASSERT_THROW(specific(shape, turns, seed), std::runtime_error)
                                << "We should not have been able to produce a difficulty this large.";
}

TYPED_TEST(ShapeShufflerTest, moderate_god_shuffle) {
    GTEST_SKIP(); /* This takes too long. */
    TypeParam shape;
    auto seed = 0;
    auto [shuffled, shuffle_path] = god(shape, seed);
    if (std::is_same<TypeParam, Volume>::value) {
        ASSERT_EQ(shuffle_path.size(), 11)
                                    << "We did not recover the god shuffle quoted in: cf: https://ruwix.com/the-rubiks-cube/gods-number/";
    } else if (std::is_same<TypeParam, Cube>::value) {
        ASSERT_EQ(shuffle_path.size(), 20)
                                    << "We did not recover the god shuffle quoted in: cf: https://ruwix.com/the-rubiks-cube/gods-number/";
    }

}

File: src/rubik/shufflers/tests/shufflers.py
#!/usr/bin/env python3
import unittest
from rubik.shufflers.shufflers import specific, god
from rubik.shapes import Domino, Volume, Square, Sheet
from common.variables import variable_names_and_objects
from common.logger import log
from common.timing import Timeout


class Shuffles(unittest.TestCase):

    def test_trivial_difficulty(self):
        for name, Shape in variable_names_and_objects(Domino, Volume):
            shape = Shape()
            log.debug(f"Testing the shape {name}")
            turns = 0
            shuffled, shuffle_path = specific(start=shape, turns=turns, seed=0)
            self.assertEqual(shuffled, shape)
            self.assertEqual(len(shuffle_path), turns)

    def test_feasible_easy_difficulty(self):
        for name, Shape in variable_names_and_objects(Domino, Volume):
            shape = Shape()
            log.debug(f"Testing the shape {name}")
            for turns in range(1, 2):
                shuffled, shuffle_path = specific(start=shape, turns=turns, seed=0)
                self.assertNotEqual(shuffled, shape)
                self.assertEqual(len(shuffle_path), turns)

    def test_feasible_moderate_difficulty(self):
        shape = Volume()
        for turns in range(1, 5):
            with Timeout(5):
                shuffled, shuffle_path = specific(start=shape, turns=turns, seed=0)
                self.assertNotEqual(shuffled, shape)
                self.assertEqual(len(shuffle_path), turns)

    def test_infeasible_moderate_difficulty(self):
        shape = Domino()
        turns = 2
        with self.assertRaises(RuntimeError, msg="We should not have been able to produce a difficulty this large."):
            shuffled, shuffle_path = specific(start=shape, turns=turns, seed=0)

    def test_easy_god_shuffle(self):
        shape = Domino()
        god_turns = 1
        with Timeout(5):
            shuffled, shuffle_path = god(start=shape, seed=0)
            self.assertNotEqual(shuffled, shape)
            self.assertEqual(len(shuffle_path), god_turns)

    ##########################################################
    # Some of the following God numbers are my best guesses...
    ##########################################################

    def test_mild_god_shuffle(self):
        shape = Square()
        god_turns = 2
        with Timeout(5):
            shuffled, shuffle_path = god(start=shape, seed=0)
            self.assertNotEqual(shuffled, shape)
            self.assertEqual(len(shuffle_path), god_turns)

    def test_moderate_god_shuffle(self):
        shape = Sheet()
        god_turns = 3
        with Timeout(5):
            shuffled, shuffle_path = god(start=shape, seed=0)
            self.assertNotEqual(shuffled, shape)
            self.assertEqual(len(shuffle_path), god_turns)

    @unittest.skip("This takes too long to compute (e.g. O(30+ mins)).")
    def test_moderate_god_shuffle(self):
        shape = Volume()
        god_turns = 11  # cf: https://ruwix.com/the-rubiks-cube/gods-number/
        with Timeout(10):
            shuffled, shuffle_path = god(start=shape, seed=0)
            self.assertNotEqual(shuffled, shape)
            self.assertEqual(len(shuffle_path), god_turns)


if __name__ == '__main__':
    from common.cli import unit_test_parse

    unit_test_parse()
    unittest.main()
